ngraph/core/include/ngraph/shape.hpp

```diff
@@ -57,6 +57,18 @@ size_t shape_size(const SHAPE_TYPE& shape) {
     return size;
 }
 
+template <typename ForwardIt>
+size_t shape_size(ForwardIt start_dim, const ForwardIt end_dim) {
+    static_assert(std::is_arithmetic<typename std::iterator_traits<ForwardIt>::value_type>::value,
+                  "shape_size expects 2 forward iterators as inputs");
+
```

Comment:
Should it throw an exception if the dimension is dynamic?

Reply:
In fact it will not compile. If you use this function with something that stores `Dimension` objects, the `static_assert` will fire because dimension is not an arithmetic type. This overload only works with simple PODs like `int` or `size_t` and that's why I added this compile-time check. I will change the error message though because it's not accurate right now.